retrieval/k8s/k8s_stats.go

```diff
@@ -1,82 +1,269 @@
+		cast.Spec.MinReadySeconds = 0
+		cast.Spec.RevisionHistoryLimit = nil
+		return cast
+	case *v1apps.ReplicaSet:
+		cast := to.(*v1apps.ReplicaSet)
+		sanitizeMeta(&cast.ObjectMeta)
+		cast.Spec.Replicas = nil
+		cast.Spec.Template = corev1.PodTemplateSpec{}
+		cast.Spec.MinReadySeconds = 0
+		return cast
+	case *v1apps.Deployment:
+		cast := to.(*v1apps.Deployment)
+		sanitizeMeta(&cast.ObjectMeta)
+		cast.Spec.Template = corev1.PodTemplateSpec{}
+		cast.Spec.Replicas = nil
+		cast.Spec.Strategy = v1apps.DeploymentStrategy{}
+		cast.Spec.MinReadySeconds = 0
+		cast.Spec.RevisionHistoryLimit = nil
+		cast.Spec.ProgressDeadlineSeconds = nil
+		return cast
+	case *v1batch.Job:
+		cast := to.(*v1batch.Job)
+		sanitizeMeta(&cast.ObjectMeta)
+		cast.Spec.Template = corev1.PodTemplateSpec{}
+		cast.Spec.Parallelism = nil
+		cast.Spec.Completions = nil
+		cast.Spec.ActiveDeadlineSeconds = nil
+		cast.Spec.BackoffLimit = nil
+		cast.Spec.ManualSelector = nil
+		cast.Spec.TTLSecondsAfterFinished = nil
+		cast.Spec.CompletionMode = nil
+		cast.Spec.Suspend = nil
+		return cast
+	case *v1batch.CronJob:
+		cast := to.(*v1batch.CronJob)
+		sanitizeMeta(&cast.ObjectMeta)
+		// cronjobs have no Selector
+		cast.Spec = v1batch.CronJobSpec{}
+		return cast
+	case *corev1.Service:
+		cast := to.(*corev1.Service)
+		sanitizeMeta(&cast.ObjectMeta)
+		cast.Spec.Ports = nil
+		cast.Spec.ClusterIP = ""
+		cast.Spec.ClusterIPs = nil
+		cast.Spec.Type = ""
+		cast.Spec.ExternalIPs = nil
+		cast.Spec.SessionAffinity = ""
+		cast.Spec.LoadBalancerIP = ""
+		cast.Spec.LoadBalancerSourceRanges = nil
+		cast.Spec.ExternalName = ""
+		cast.Spec.ExternalTrafficPolicy = ""
+		cast.Spec.HealthCheckNodePort = 0
+		cast.Spec.SessionAffinityConfig = nil
+		cast.Spec.IPFamilies = nil
+		cast.Spec.IPFamilyPolicy = nil
+		cast.Spec.AllocateLoadBalancerNodePorts = nil
+		cast.Spec.LoadBalancerClass = nil
+		cast.Spec.InternalTrafficPolicy = nil
+		return cast
+	case *corev1.ReplicationController:
+		cast := to.(*corev1.ReplicationController)
+		sanitizeMeta(&cast.ObjectMeta)
+		cast.Spec.Replicas = nil
+		cast.Spec.Template = nil
+		cast.Spec.MinReadySeconds = 0
+		return cast
+	case *corev1.Namespace:
+		return sanitizeNamespace(to)
+	case *corev1.PersistentVolume:
+		cast := to.(*corev1.PersistentVolume)
+		sanitizeMeta(&cast.ObjectMeta)
+		return cast
+	case *corev1.PersistentVolumeClaim:
+		cast := to.(*corev1.PersistentVolumeClaim)
+		sanitizeMeta(&cast.ObjectMeta)
+		return cast
+	case *corev1.Node:
+		cast := to.(*corev1.Node)
+		sanitizeMeta(&cast.ObjectMeta)
+		return cast
 	}
-	v1Sources = []string{
-		"namespaces",
-		"replicationcontrollers",
-		"services",
-		"nodes",
-		"pods",
-		"persistentvolumes",
-		"persistentvolumeclaims",
+	return to
+}
+
+func sanitizeMeta(objectMeta *metav1.ObjectMeta) {
+	objectMeta.ManagedFields = nil
+	if _, ok := objectMeta.Annotations[KubernetesLastAppliedConfig]; ok {
+		delete(objectMeta.Annotations, KubernetesLastAppliedConfig)
```

Comment:
Delete is a noop if the key doesn't exist, safe to call without the check.

Reply:
I took this logic from your first implementation, does that mean I can also remove if's for sanitizePod's delete calls?

Reply:
Yep. My bad for putting it there to begin with.

Reply:
np just wanted to check. removed the if check